docs/benchmark.md

```diff
@@ -0,0 +1,73 @@
+# Benchmark
+
+Go benchmark tests that measure end-to-end performance of a running Ollama server. Run these tests to evaluate model inference performance on your hardware and measure the impact of code changes.
+
+## When to use
+
+Run these benchmarks when:
+- Making changes to the model inference engine
+- Modifying model loading/unloading logic
+- Changing prompt processing or token generation code
+- Implementing a new model architecture
+- Testing performance across different hardware setups
+
+## Prerequisites
+- Ollama server running locally (`127.0.0.1:11434`)
+- benchstat tool: `go install golang.org/x/perf/cmd/benchstat@latest`
+
+## Usage and Examples
+
+>[!NOTE]
+>All commands must be run from the `benchmark` directory of the Ollama project.
+
+Basic syntax:
+```bash
+go test -bench=. -m MODEL_NAME ./...
```

Comment:
Probably need to pull a model first? How would that work? I guess just an `ollama pull` ahead of time?

Reply:
Yeah, you will need an `ollama pull ...` or import ahead of time. I don't want to be too perscriptive here since this benchmark will often be used while integrating a new model, before its available to pull.

For reference, this is what it looks like when you try to run the benchmark with a model you do not have:
```
❯ go test -bench=. ./benchmark/... -m dne

--- FAIL: BenchmarkColdStart
    server_benchmark_test.go:146: Model unavailable: model 'dne' not found
--- FAIL: BenchmarkWarmStart
    server_benchmark_test.go:146: Model unavailable: model 'dne' not found
FAIL
exit status 1
FAIL    github.com/ollama/ollama/benchmark      0.234s
FAIL
```

Reply:
Ok that's fine 👍